src/main/scala/rules/Joiner.scala

```diff
@@ -40,15 +46,25 @@ object joiner extends JoiningRules {
       val s2 = s1.copy(underJoin = true)
 
       block(s2, v1, (s3, data, v2) => {
-        /* In order to prevent mismatches between different final states of the evaluation
-         * paths that are to be joined, we reset certain state properties that may have been
-         * affected by the evaluation - such as the store (by let-bindings) or the heap (by
-         * state consolidations) to their initial values.
-         */
-        val s4 = s3.copy(g = s1.g,
-                         h = s1.h,
-                         oldHeaps = s1.oldHeaps,
-                         underJoin = s1.underJoin)
+        val s4 =
+          if (resetState) {
+            /* In order to prevent mismatches between different final states of the evaluation
+             * paths that are to be joined, we reset certain state properties that may have been
+             * affected by the evaluation - such as the store (by let-bindings) or the heap (by
+             * state consolidations) to their initial values.
+             */
+            s3.copy(g = s1.g,
+                    h = s1.h,
+                    oldHeaps = s1.oldHeaps,
+                    underJoin = s1.underJoin,
+                    // TODO: Evaluation should not affect partiallyConsumedHeap, probably
+                    ssCache = s1.ssCache,
+                    partiallyConsumedHeap = s1.partiallyConsumedHeap,
+                    invariantContexts = s1.invariantContexts)
+          } else {
+            // For more joins, state shouldn't be reset.
```

Comment:
Does this express the assumption that `resetState == Config.moreJoins()`? If so, why is `resetState` needed?

Reply:
No, `resetState` can be true or false if `moreJoins`is enabled. However, `moreJoins == false ==> resetState == true`.

Intuitively, `resetState` is true for "pure" branches which already are joined without `moreJoins`, and false for impure branches, which are only joined with `moreJoins` enabled.